illwill.nim

```diff
@@ -658,53 +629,52 @@ else:  # OS X & Linux
     else:
       gMouseInfo.scrollDir = ScrollDirection.sdNone
 
-  proc parseKeys(charsRead: int): seq[Key] =
-    # Inspired by
-    # https://github.com/mcandre/charm/blob/master/lib/charm.c
-    if charsRead == 1:
-      let ch = keyBuf[0]
-      case ch:
-      of   9: result.add(Key.Tab)
-      of  10: result.add(Key.Enter)
-      of  27: result.add(Key.Escape)
-      of  32: result.add(Key.Space)
-      of 127: result.add(Key.Backspace)
-      of 0, 29, 30, 31: discard   # these have no Windows equivalents so
-                                  # we'll ignore them
+  proc parseStdin[T](input: T): Key =
+    var ch1, ch2, ch3, ch4, ch5: char
+    result = Key.None
+    if read(input, ch1.addr, 1) > 0:
+      case ch1
+      of '\e':
+        if read(input, ch2.addr, 1) > 0:
+          if ch2 == 'O' and read(input, ch3.addr, 1) > 0:
+            if ch3 in "ABCDFH":
+              result = KEYS_D[int(ch3) - int('A')]
+            elif ch3 in "PQRS":
+              result = KEYS_F[int(ch3) - int('P')]
+          elif ch2 == '[' and read(input, ch3.addr, 1) > 0:
+            if ch3 in "ABCDFH":
+              result = KEYS_D[int(ch3) - int('A')]
+            elif ch3 in "PQRS":
+              result = KEYS_F[int(ch3) - int('P')]
+            elif ch3 == '1' and read(input, ch4.addr, 1) > 0:
+              if ch4 == '~':
+                result = Key.Home
+              elif ch4 in "12345789" and read(input, ch5.addr, 1) > 0 and ch5 == '~':
+                result = KEYS_F[int(ch4) - int('1')]
+            elif ch3 == '2' and read(input, ch4.addr, 1) > 0:
+              if ch4 == '~':
+                result = Key.Insert
+              elif ch4 in "0134" and read(input, ch5.addr, 1) > 0 and ch5 == '~':
+                result = KEYS_G[int(ch4) - int('0')]
+            elif ch3 in "345678" and read(input, ch4.addr, 1) > 0 and ch4 == '~':
+              result = KEYS_E[int(ch3) - int('3')]
+            else:
```

Comment:
Okay, I'm _really_ not looking forward to maintaining this... Can you make this buffered, or table-based, or anything that makes it not look like the dog's breakfast? 😆 Please optimise for readability.

Reply:
I did not find to add any additional buffers and to create additional logic to maintain it because stdin is already buffer. Yep, the code is pretty ugly, but it is not the first implementation and much better than the first one

Reply:
The reason why I added the tests - to support the complicate function